File: src-tauri/src/database/commands.rs
use std::sync::Arc;
use uuid::Uuid;
use tauri::State;
use serde_json::value::RawValue;

use crate::{
    database::{
        services::{
            connection::ConnectionService,
            query::QueryService,
            mutation::MutationService,
            metadata::MetadataService,
        },
        types::{
            ConnectionInfo, DatabaseInfo, StatementInfo, QueryStatus,
            DatabaseSchema,
        },
        maintenance::{SoftDeleteResult, TruncateResult, DumpResult},
    },
    error::Error,
    AppState,
    storage::{QueryHistoryEntry, SavedQuery, ConnectionHistoryEntry},
};

// Re-exports for compatibility and accessibility
pub use crate::database::services::mutation::{ExportFormat, MutationResult};
pub use crate::database::metadata::DatabaseMetadata;

// Re-export helpers for maintenance.rs
pub use crate::database::services::mutation::{json_to_pg_param, json_to_sqlite_value};

// =============================================================================
// Connection Commands
// =============================================================================

#[tauri::command]
pub async fn add_connection(
    name: String,
    database_info: DatabaseInfo,
    color: Option<i32>,
    state: State<'_, AppState>,
) -> Result<ConnectionInfo, Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.add_connection(name, database_info, color).await
}

#[tauri::command]
pub async fn update_connection(
    conn_id: Uuid,
    name: String,
    database_info: DatabaseInfo,
    color: Option<i32>,
    state: State<'_, AppState>,
) -> Result<ConnectionInfo, Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.update_connection(conn_id, name, database_info, color).await
}

#[tauri::command]
pub async fn update_connection_color(
    connection_id: Uuid,
    color: Option<i32>,
    state: State<'_, AppState>,
) -> Result<(), Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.update_connection_color(connection_id, color).await
}

#[tauri::command]
pub async fn connect_to_database(
    connection_id: Uuid,
    state: State<'_, AppState>,
    monitor: State<'_, crate::database::ConnectionMonitor>,
    certificates: State<'_, crate::database::Certificates>,
) -> Result<bool, Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.connect_to_database(&monitor, &certificates, connection_id).await
}

#[tauri::command]
pub async fn disconnect_from_database(
    connection_id: Uuid,
    state: State<'_, AppState>,
) -> Result<(), Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.disconnect_from_database(connection_id).await
}

#[tauri::command]
pub async fn get_connections(state: State<'_, AppState>) -> Result<Vec<ConnectionInfo>, Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.get_connections().await
}

#[tauri::command]
pub async fn remove_connection(
    connection_id: Uuid,
    state: State<'_, AppState>,
) -> Result<(), Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.remove_connection(connection_id).await
}

#[tauri::command]
pub async fn test_connection(
    database_info: DatabaseInfo,
    certificates: State<'_, crate::database::Certificates>,
) -> Result<bool, Error> {
    ConnectionService::test_connection(database_info, &certificates).await
}

#[tauri::command]
pub async fn initialize_connections(state: State<'_, AppState>) -> Result<(), Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.initialize_connections().await
}

#[tauri::command]
pub async fn get_recent_connections(
    limit: Option<u32>,
    state: State<'_, AppState>,
) -> Result<Vec<ConnectionInfo>, Error> {
    let svc = ConnectionService {
        connections: &state.connections,
        storage: &state.storage,
    };
    svc.get_recent_connections(limit).await
}

#[tauri::command]
pub async fn get_connection_history(
    db_type_filter: Option<String>,
    success_filter: Option<bool>,
    limit: Option<i64>,
    state: State<'_, AppState>,
) -> Result<Vec<ConnectionHistoryEntry>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_connection_history(db_type_filter, success_filter, limit).await
}


// =============================================================================
// Query Commands
// =============================================================================

#[tauri::command]
pub async fn start_query(
    connection_id: Uuid,
    query: &str,
    state: State<'_, AppState>,
) -> Result<Vec<usize>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.start_query(connection_id, query).await
}

#[tauri::command]
pub async fn fetch_query(
    query_id: usize,
    state: State<'_, AppState>,
) -> Result<StatementInfo, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.fetch_query(query_id).await
}

#[tauri::command]
pub async fn fetch_page(
    query_id: usize,
    page_index: usize,
    state: State<'_, AppState>,
) -> Result<Option<Box<RawValue>>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.fetch_page(query_id, page_index).await
}

#[tauri::command]
pub async fn get_query_status(
    query_id: usize,
    state: State<'_, AppState>,
) -> Result<QueryStatus, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_query_status(query_id).await
}

#[tauri::command]
pub async fn get_page_count(
    query_id: usize,
    state: State<'_, AppState>,
) -> Result<usize, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_page_count(query_id).await
}

#[tauri::command]
pub async fn get_columns(
    query_id: usize,
    state: State<'_, AppState>,
) -> Result<Option<Box<RawValue>>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_columns(query_id).await
}

#[tauri::command]
pub async fn save_query_to_history(
    connection_id: String,
    query: String,
    duration_ms: Option<u64>,
    status: String,
    row_count: u64,
    error_message: Option<String>,
    state: State<'_, AppState>,
) -> Result<(), Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.save_query_to_history(connection_id, query, duration_ms, status, row_count, error_message).await
}

#[tauri::command]
pub async fn get_query_history(
    connection_id: String,
    limit: Option<u32>,
    state: State<'_, AppState>,
) -> Result<Vec<QueryHistoryEntry>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_query_history(connection_id, limit).await
}

#[tauri::command]
pub async fn get_recent_queries(
    connection_id: Option<String>,
    limit: Option<u32>,
    status_filter: Option<String>,
    state: State<'_, AppState>,
) -> Result<Vec<QueryHistoryEntry>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_recent_queries(connection_id, limit, status_filter).await
}


// =============================================================================
// Script Commands
// =============================================================================

#[tauri::command]
pub async fn save_script(
    name: String,
    content: String,
    connection_id: Option<Uuid>,
    description: Option<String>,
    state: State<'_, AppState>,
) -> Result<i64, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.save_script(name, content, connection_id, description).await
}

#[tauri::command]
pub async fn update_script(
    id: i64,
    name: String,
    content: String,
    connection_id: Option<Uuid>,
    description: Option<String>,
    state: State<'_, AppState>,
) -> Result<(), Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.update_script(id, name, content, connection_id, description).await
}

#[tauri::command]
pub async fn get_scripts(
    connection_id: Option<Uuid>,
    state: State<'_, AppState>,
) -> Result<Vec<SavedQuery>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_scripts(connection_id).await
}

#[tauri::command]
pub async fn delete_script(id: i64, state: State<'_, AppState>) -> Result<(), Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.delete_script(id).await
}


// =============================================================================
// Setting / Session Commands
// =============================================================================

#[tauri::command]
pub async fn save_session_state(
    session_data: &str,
    state: State<'_, AppState>,
) -> Result<(), Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.save_session_state(session_data).await
}

#[tauri::command]
pub async fn get_session_state(state: State<'_, AppState>) -> Result<Option<String>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_session_state().await
}

#[tauri::command]
pub async fn get_setting(key: String, state: State<'_, AppState>) -> Result<Option<String>, Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.get_setting(key).await
}

#[tauri::command]
pub async fn set_setting(
    key: String,
    value: String,
    state: State<'_, AppState>,
) -> Result<(), Error> {
    let svc = QueryService {
        connections: &state.connections,
        storage: &state.storage,
        stmt_manager: &state.stmt_manager,
    };
    svc.set_setting(key, value).await
}


// =============================================================================
// Mutation Commands
// =============================================================================

#[tauri::command]
pub async fn insert_row(
    connection_id: Uuid,
    table_name: String,
    schema_name: Option<String>,
    row_data: serde_json::Map<String, serde_json::Value>,
    state: State<'_, AppState>,
) -> Result<MutationResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.insert_row(connection_id, table_name, schema_name, row_data).await
}

#[tauri::command]
pub async fn update_cell(
    connection_id: Uuid,
    table_name: String,
    schema_name: Option<String>,
    primary_key_column: String,
    primary_key_value: serde_json::Value,
    column_name: String,
    new_value: serde_json::Value,
    state: State<'_, AppState>,
) -> Result<MutationResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.update_cell(connection_id, table_name, schema_name, primary_key_column, primary_key_value, column_name, new_value).await
}

#[tauri::command]
pub async fn delete_rows(
    connection_id: Uuid,
    table_name: String,
    schema_name: Option<String>,
    primary_key_column: String,
    primary_key_values: Vec<serde_json::Value>,
    state: State<'_, AppState>,
) -> Result<MutationResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.delete_rows(connection_id, table_name, schema_name, primary_key_column, primary_key_values).await
}

#[tauri::command]
pub async fn export_table(
    connection_id: Uuid,
    table_name: String,
    schema_name: Option<String>,
    format: ExportFormat,
    limit: Option<u32>,
    state: State<'_, AppState>,
) -> Result<String, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.export_table(connection_id, table_name, schema_name, format, limit).await
}

#[tauri::command]
pub async fn soft_delete_rows(
    connection_id: Uuid,
    table_name: String,
    schema_name: Option<String>,
    primary_key_column: String,
    primary_key_values: Vec<serde_json::Value>,
    soft_delete_column: Option<String>,
    state: State<'_, AppState>,
) -> Result<SoftDeleteResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.soft_delete_rows(connection_id, table_name, schema_name, primary_key_column, primary_key_values, soft_delete_column).await
}

#[tauri::command]
pub async fn undo_soft_delete(
    connection_id: Uuid,
    table_name: String,
    schema_name: Option<String>,
    primary_key_column: String,
    primary_key_values: Vec<serde_json::Value>,
    soft_delete_column: String,
    state: State<'_, AppState>,
) -> Result<MutationResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.undo_soft_delete(connection_id, table_name, schema_name, primary_key_column, primary_key_values, soft_delete_column).await
}

#[tauri::command]
pub async fn truncate_table(
    connection_id: Uuid,
    table_name: String,
    schema_name: Option<String>,
    cascade: Option<bool>,
    state: State<'_, AppState>,
) -> Result<TruncateResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.truncate_table(connection_id, table_name, schema_name, cascade).await
}

#[tauri::command]
pub async fn truncate_database(
    connection_id: Uuid,
    schema_name: Option<String>,
    confirm: bool,
    state: State<'_, AppState>,
) -> Result<TruncateResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.truncate_database(connection_id, schema_name, confirm).await
}

#[tauri::command]
pub async fn dump_database(
    connection_id: Uuid,
    output_path: String,
    state: State<'_, AppState>,
) -> Result<DumpResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.dump_database(connection_id, output_path).await
}

#[tauri::command]
pub async fn execute_batch(
    connection_id: Uuid,
    statements: Vec<String>,
    state: State<'_, AppState>,
) -> Result<MutationResult, Error> {
    let svc = MutationService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.execute_batch(connection_id, statements).await
}


// =============================================================================
// Metadata Commands
// =============================================================================

#[tauri::command]
pub async fn get_database_schema(
    connection_id: Uuid,
    state: State<'_, AppState>,
) -> Result<Arc<DatabaseSchema>, Error> {
    let svc = MetadataService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.get_database_schema(connection_id).await
}

#[tauri::command]
pub async fn get_database_metadata(
    connection_id: Uuid,
    state: State<'_, AppState>,
) -> Result<DatabaseMetadata, Error> {
    let svc = MetadataService {
        connections: &state.connections,
        schemas: &state.schemas,
    };
    svc.get_database_metadata(connection_id).await
}

File: _old/src/core/tauri/commands.ts
import { invoke } from '@tauri-apps/api/core'
import type {
  ConnectionInfo,
  DatabaseInfo,
  Schema,
  QueryHistoryEntry,
  Script,
  StatementInfo,
  QueryId,
  QueryStatus,
  Page,
} from '@/types/database'

export async function testConnection(databaseInfo: DatabaseInfo): Promise<boolean> {
  return await invoke('test_connection', { databaseInfo })
}

export async function addConnection(
  name: string,
  databaseInfo: DatabaseInfo
): Promise<ConnectionInfo> {
  return await invoke('add_connection', { name, databaseInfo })
}

export async function connectToDatabase(connectionId: string): Promise<boolean> {
  return await invoke('connect_to_database', { connectionId })
}

export async function disconnectFromDatabase(connectionId: string): Promise<void> {
  return await invoke('disconnect_from_database', { connectionId })
}

export async function getConnections(): Promise<ConnectionInfo[]> {
  return await invoke('get_connections')
}

export async function removeConnection(connectionId: string): Promise<void> {
  return await invoke('remove_connection', { connectionId })
}

export async function updateConnection(
  connectionId: string,
  name: string,
  databaseInfo: DatabaseInfo
): Promise<ConnectionInfo> {
  return await invoke('update_connection', { connId: connectionId, name, databaseInfo })
}

export async function initializeConnections(): Promise<void> {
  return await invoke('initialize_connections')
}

export async function saveQueryToHistory(
  connectionId: string,
  query: string,
  durationMs?: number,
  status: string = 'success',
  rowCount: number = 0,
  errorMessage?: string
): Promise<void> {
  await invoke('save_query_to_history', {
    connectionId,
    query,
    durationMs,
    status,
    rowCount,
    errorMessage,
  })
}

export async function getQueryHistory(
  connectionId: string,
  limit?: number
): Promise<QueryHistoryEntry[]> {
  return await invoke('get_query_history', { connectionId, limit })
}

export async function getDatabaseSchema(connectionId: string): Promise<Schema> {
  return await invoke('get_database_schema', { connectionId })
}

export async function getSchema(connectionId: string): Promise<Schema> {
  return await invoke('get_database_schema', { connectionId })
}

export async function refreshSchema(connectionId: string): Promise<Schema> {
  return await invoke('get_database_schema', { connectionId })
}

export async function saveScript(
  name: string,
  content: string,
  connectionId?: string,
  description?: string
): Promise<number> {
  return await invoke('save_script', {
    name,
    content,
    connectionId: connectionId || null,
    description: description || null,
  })
}

export async function updateScript(
  id: number,
  name: string,
  content: string,
  connectionId?: string,
  description?: string
): Promise<void> {
  return await invoke('update_script', {
    id,
    name,
    content,
    connectionId: connectionId || null,
    description: description || null,
  })
}

export async function getScripts(connectionId?: string): Promise<Script[]> {
  return await invoke('get_scripts', { connectionId: connectionId || null })
}

export async function deleteScript(id: number): Promise<void> {
  await invoke('delete_script', { id })
}

export async function minimizeWindow(): Promise<void> {
  await invoke('minimize_window')
}

export async function maximizeWindow(): Promise<void> {
  await invoke('maximize_window')
}

export async function closeWindow(): Promise<void> {
  await invoke('close_window')
}

export async function saveSessionState(sessionData: string): Promise<void> {
  return await invoke('save_session_state', { sessionData })
}

export async function getSessionState(): Promise<string | null> {
  return await invoke('get_session_state')
}

export async function getSetting<T = string>(key: string): Promise<T | null> {
  const value = await invoke<string | null>('get_setting', { key })
  if (value === null) return null
  try {
    return JSON.parse(value) as T
  } catch {
    return value as unknown as T
  }
}

export async function setSetting(key: string, value: any): Promise<void> {
  const stringValue = typeof value === 'string' ? value : JSON.stringify(value)
  return await invoke('set_setting', { key, value: stringValue })
}

export async function pickSqliteDbDialog(): Promise<string | null> {
  return await invoke('open_sqlite_db')
}

export async function saveSqliteDbDialog(): Promise<string | null> {
  return await invoke('save_sqlite_db')
}

export async function startQuery(connectionId: string, query: string): Promise<QueryId[]> {
  return await invoke('start_query', { connectionId, query })
}

export async function fetchQuery(queryId: QueryId): Promise<StatementInfo> {
  return await invoke('fetch_query', { queryId })
}

export async function fetchPage(queryId: QueryId, pageIndex: number): Promise<Page | null> {
  return await invoke('fetch_page', { queryId, pageIndex })
}

export async function getQueryStatus(queryId: QueryId): Promise<QueryStatus> {
  return await invoke('get_query_status', { queryId })
}

export async function getPageCount(queryId: QueryId): Promise<number> {
  return await invoke('get_page_count', { queryId })
}

export async function getColumns(queryId: QueryId): Promise<string[] | null> {
  return await invoke('get_columns', { queryId })
}

export async function executeQuery(
  connectionId: string,
  query: string
): Promise<StatementInfo[]> {
  const queryIds = await startQuery(connectionId, query)
  const results: StatementInfo[] = []

  for (const queryId of queryIds) {
    const result = await fetchQuery(queryId)
    results.push(result)
  }

  return results
}

export async function getConnectionHistory(
  dbTypeFilter?: string,
  successFilter?: boolean,
  limit?: number
): Promise<import('@/types/database').ConnectionHistoryEntry[]> {
  return await invoke('get_connection_history', {
    dbTypeFilter: dbTypeFilter || null,
    successFilter: successFilter ?? null,
    limit: limit || null,
  })
}

export async function populateTestQueries(): Promise<string> {
  return await invoke('populate_test_queries_command')
}

// =============================================================================
// Mutation API Commands
// =============================================================================
// These commands provide structured data manipulation for spreadsheet-style UIs

/** Export format options for table data */
export type ExportFormat = 'json' | 'sql_insert' | 'csv'

/** Result of a mutation operation */
export type MutationResult = {
  success: boolean
  affected_rows: number
  message: string | null
}

/**
 * Update a single cell value in a table
 * @param connectionId - UUID of the database connection
 * @param tableName - Name of the table to update
 * @param schemaName - Schema name (optional, for PostgreSQL)
 * @param primaryKeyColumn - Name of the primary key column
 * @param primaryKeyValue - Value of the primary key for the row to update
 * @param columnName - Name of the column to update
 * @param newValue - New value to set
 */
export async function updateCell(
  connectionId: string,
  tableName: string,
  schemaName: string | null,
  primaryKeyColumn: string,
  primaryKeyValue: unknown,
  columnName: string,
  newValue: unknown
): Promise<MutationResult> {
  return await invoke('update_cell', {
    connectionId,
    tableName,
    schemaName,
    primaryKeyColumn,
    primaryKeyValue,
    columnName,
    newValue,
  })
}

/**
 * Delete one or more rows from a table
 * @param connectionId - UUID of the database connection
 * @param tableName - Name of the table
 * @param schemaName - Schema name (optional, for PostgreSQL)
 * @param primaryKeyColumn - Name of the primary key column
 * @param primaryKeyValues - Array of primary key values for rows to delete
 */
export async function deleteRows(
  connectionId: string,
  tableName: string,
  schemaName: string | null,
  primaryKeyColumn: string,
  primaryKeyValues: unknown[]
): Promise<MutationResult> {
  return await invoke('delete_rows', {
    connectionId,
    tableName,
    schemaName,
    primaryKeyColumn,
    primaryKeyValues,
  })
}

/**
 * Export table data to a specific format
 * @param connectionId - UUID of the database connection
 * @param tableName - Name of the table to export
 * @param schemaName - Schema name (optional, for PostgreSQL)
 * @param format - Export format: 'json', 'sql_insert', or 'csv'
 * @param limit - Optional row limit
 */
export async function exportTable(
  connectionId: string,
  tableName: string,
  schemaName: string | null,
  format: ExportFormat,
  limit?: number
): Promise<string> {
  return await invoke('export_table', {
    connectionId,
    tableName,
    schemaName,
    format,
    limit: limit ?? null,
  })
}

// =============================================================================
// Database Metadata Types & Commands
// =============================================================================

/** Database-level metadata information */
export type DatabaseMetadata = {
  size_bytes: number
  created_at: number | null
  last_updated: number | null
  row_count_total: number
  table_count: number
  host: string
  database_name: string | null
}

/**
 * Get database-level metadata (size, row counts, dates, etc.)
 * @param connectionId - UUID of the database connection
 */
export async function getDatabaseMetadata(connectionId: string): Promise<DatabaseMetadata> {
  return await invoke('get_database_metadata', { connectionId })
}

// =============================================================================
// Soft Delete Types & Commands
// =============================================================================

/** Result of a soft delete operation */
export type SoftDeleteResult = {
  success: boolean
  affected_rows: number
  message: string | null
  deleted_at: number
  undo_window_seconds: number
}

/**
 * Soft delete rows (set deleted_at timestamp instead of hard delete)
 * @param connectionId - UUID of the database connection
 * @param tableName - Name of the table
 * @param schemaName - Schema name (optional)
 * @param primaryKeyColumn - Name of the primary key column
 * @param primaryKeyValues - Array of primary key values for rows to soft delete
 * @param softDeleteColumn - Name of the soft delete column (auto-detected if not provided)
 */
export async function softDeleteRows(
  connectionId: string,
  tableName: string,
  schemaName: string | null,
  primaryKeyColumn: string,
  primaryKeyValues: unknown[],
  softDeleteColumn?: string
): Promise<SoftDeleteResult> {
  return await invoke('soft_delete_rows', {
    connectionId,
    tableName,
    schemaName,
    primaryKeyColumn,
    primaryKeyValues,
    softDeleteColumn: softDeleteColumn ?? null,
  })
}

/**
 * Undo a soft delete (restore rows)
 * @param connectionId - UUID of the database connection
 * @param tableName - Name of the table
 * @param schemaName - Schema name (optional)
 * @param primaryKeyColumn - Name of the primary key column
 * @param primaryKeyValues - Array of primary key values for rows to restore
 * @param softDeleteColumn - Name of the soft delete column
 */
export async function undoSoftDelete(
  connectionId: string,
  tableName: string,
  schemaName: string | null,
  primaryKeyColumn: string,
  primaryKeyValues: unknown[],
  softDeleteColumn: string
): Promise<MutationResult> {
  return await invoke('undo_soft_delete', {
    connectionId,
    tableName,
    schemaName,
    primaryKeyColumn,
    primaryKeyValues,
    softDeleteColumn,
  })
}

// =============================================================================
// Truncate Types & Commands
// =============================================================================

/** Result of a truncate operation */
export type TruncateResult = {
  success: boolean
  affected_rows: number
  tables_truncated: string[]
  message: string | null
}

/**
 * Truncate a single table
 * @param connectionId - UUID of the database connection
 * @param tableName - Name of the table to truncate
 * @param schemaName - Schema name (optional)
 * @param cascade - Whether to cascade truncate to related tables (PostgreSQL only)
 */
export async function truncateTable(
  connectionId: string,
  tableName: string,
  schemaName: string | null,
  cascade?: boolean
): Promise<TruncateResult> {
  return await invoke('truncate_table', {
    connectionId,
    tableName,
    schemaName,
    cascade: cascade ?? false,
  })
}

/**
 * Truncate all tables in the database (DANGEROUS - requires confirmation)
 * @param connectionId - UUID of the database connection
 * @param schemaName - Schema name (optional, defaults to 'public' for PostgreSQL)
 * @param confirm - Must be true to execute (safety check)
 */
export async function truncateDatabase(
  connectionId: string,
  schemaName: string | null,
  confirm: boolean
): Promise<TruncateResult> {
  return await invoke('truncate_database', {
    connectionId,
    schemaName,
    confirm,
  })
}

// =============================================================================
// Database Dump Types & Commands
// =============================================================================

/** Result of a database dump operation */
export type DumpResult = {
  success: boolean
  file_path: string
  size_bytes: number
  tables_dumped: number
  rows_dumped: number
  message: string | null
}

/**
 * Dump database to a SQL file
 * @param connectionId - UUID of the database connection
 * @param outputPath - Path where to save the dump file
 */
export async function dumpDatabase(
  connectionId: string,
  outputPath: string
): Promise<DumpResult> {
  return await invoke('dump_database', {
    connectionId,
    outputPath,
  })
}

// =============================================================================
// Recent Queries & Connections Commands
// =============================================================================

/**
 * Get recent queries with enhanced filtering
 * @param connectionId - Optional connection ID to filter by
 * @param limit - Maximum number of results
 * @param statusFilter - Optional status filter ('success', 'error', etc.)
 */
export async function getRecentQueries(
  connectionId?: string,
  limit?: number,
  statusFilter?: string
): Promise<QueryHistoryEntry[]> {
  return await invoke('get_recent_queries', {
    connectionId: connectionId ?? null,
    limit: limit ?? null,
    statusFilter: statusFilter ?? null,
  })
}

/**
 * Get recent connections (last connected databases)
 * @param limit - Maximum number of results
 */
export async function getRecentConnections(
  limit?: number
): Promise<ConnectionInfo[]> {
  return await invoke('get_recent_connections', {
    limit: limit ?? null,
  })
}

